File: 599-minimum-index-sum-of-two-lists/minimum-index-sum-of-two-lists.cpp
class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        unordered_map<string,vector<int>> words;
        for(int i=0;i<list1.size();++i) words[list1[i]].push_back(i);
        for(int i=0;i<list2.size();++i) words[list2[i]].push_back(i);
        int sum=INT_MAX;
        vector<string> sol;
        for(const auto& [word,indices]:words){
            if(indices.size()>1){
                int val=0;
                for(int idx:indices) val+=idx;
                if(val<sum){
                    sol.clear();
                    sum=val;
                    sol.push_back(word);
                }
                else if(val==sum) sol.push_back(word);
            } 
        }
        return sol;
    }
};

File: 599-minimum-index-sum-of-two-lists/minimum-index-sum-of-two-lists.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        str_sum = {}
        sol_str_sum={}

        for idx,string in enumerate(list1):
            str_sum[string] = idx
        
        for idx,string in enumerate(list2):
            if string in str_sum:
                sol_str_sum[string] = idx + str_sum[string]

        print(sol_str_sum)
            
        smallest = 1e9
        sol = []
        for k,v in sol_str_sum.items():
            if v<smallest:
                sol = [k]
                smallest = v
            elif v==smallest:
                sol.append(k)
        
        return sol